### PR TITLE
Feature: Added TokenMetadata table

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**

### CODE REVIEW

File: dbschema/1.tables.sql [hunk 1]
```diff
@@ -20,3 +20,22 @@ CREATE INDEX tbl_token_transfers_value ON tbl_token_transfers (value);
 CREATE INDEX tbl_token_transfers_block_date ON tbl_token_transfers (block_date);
 CREATE INDEX tbl_token_transfers_block_number ON tbl_token_transfers (block_number);
 CREATE INDEX tbl_token_transfers_block_hash ON tbl_token_transfers (block_hash);
+
+CREATE TABLE tbl_token_metadata (
```
Comment (krishan711) [inline dbschema/1.tables.sql:24]: table name should be plural. I would actually just call this `tbl_tokens` I think

File: dbschema/1.tables.sql [hunk 2]
```diff
@@ -20,3 +20,22 @@ CREATE INDEX tbl_token_transfers_value ON tbl_token_transfers (value);
...
+    created_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    updated_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    registry_address TEXT NOT NULL,
+    token_id TEXT NOT NULL,
+    metadata_url TEXT NOT NULL,
+    image_url TEXT NOT NULL
+    name TEXT NOT NULL,
+    description TEXT NOT NULL,
+    attributes TEXT NOT NULL
+
```
Comment (krishan711) [inline dbschema/1.tables.sql:35]: remove these empty lines

File: dbschema/1.tables.sql [hunk 3]
```diff
@@ -20,3 +20,22 @@ CREATE INDEX tbl_token_transfers_value ON tbl_token_transfers (value);
...
 CREATE INDEX tbl_token_transfers_block_hash ON tbl_token_transfers (block_hash);
+
+CREATE TABLE tbl_token_metadata (
+    id BIGSERIAL PRIMARY KEY,
+    created_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    updated_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    registry_address TEXT NOT NULL,
+    token_id TEXT NOT NULL,
+    metadata_url TEXT NOT NULL,
+    image_url TEXT NOT NULL
```
Comment (krishan711) [inline dbschema/1.tables.sql:31]: are you sure this should be NOT NULL?

File: dbschema/1.tables.sql [hunk 4]
```diff
@@ -20,3 +20,22 @@ CREATE INDEX tbl_token_transfers_value ON tbl_token_transfers (value);
...
+    name TEXT NOT NULL,
+    description TEXT NOT NULL,
```
Comment (krishan711) [inline dbschema/1.tables.sql:33]: also sure this should be NOT NULL?

File: dbschema/1.tables.sql [hunk 5]
```diff
@@ -20,3 +20,22 @@ CREATE INDEX tbl_token_transfers_value ON tbl_token_transfers (value);
...
+    attributes TEXT NOT NULL
+
+
+);
+CREATE UNIQUE INDEX tbl_token_registry_address_token_id ON tbl_token_metadata (registry_address, token_id);
```
Comment (krishan711) [inline dbschema/1.tables.sql:38]: the names of these indices (the part after CREATE UNIQUE INDEX) should follow a formault of `<table_name>_<field_name_1>_<field_name_2>_...`

File: api/notd/store/saver.py [hunk 6]
```diff
@@ -51,3 +52,28 @@ async def create_token_transfer(self, retrievedTokenTransfer: RetrievedTokenTran
             blockHash=retrievedTokenTransfer.blockHash,
             blockDate=retrievedTokenTransfer.blockDate,
         )
+    async def create_offchain_content(self, tokenId: int, registryAddress: str, metadataUrl: str, imageUrl: int, name: str, description: str, attributes: str) -> TokenMetadata:
```
Comment (krishan711) [inline api/notd/store/saver.py:55]: fix name

File: api/notd/model.py [hunk 7]
```diff
@@ -20,6 +20,18 @@ class RetrievedTokenTransfer:
     blockHash: str
     blockDate: datetime.datetime
 
+@dataclasses.dataclass
+class TokenMetadata:
```
Comment (krishan711) [inline api/notd/model.py:24]: would just call this `Token`

File: api/notd/store/retriever.py [hunk 8]
```diff
@@ -9,10 +9,10 @@
 from core.store.retriever import StringFieldFilter
 from sqlalchemy.sql.expression import func as sqlalchemyfunc
 
-from notd.model import Token
+from notd.model import Token, TokenMetadata
 from notd.model import TokenTransfer
-from notd.store.schema import TokenTransfersTable
-from notd.store.schema_conversions import token_transfer_from_row
+from notd.store.schema import TokenTransfersTable,TokenMetadataTable
+from notd.store.schema_conversions import token_transfer_from_row,token_metadata_from_row
```
Comment (krishan711) [inline api/notd/store/retriever.py:15]: put new imports on new lines pls. you can run `isort --sl notd` in the api/ folder and it should do it for you

File: api/notd/store/saver.py (hunk 6 above)
Comment (krishan711) [reply api/notd/store/saver.py:55]: also newline between functions

Comment (krishan711) [reply api/notd/store/saver.py:55]: also pls add an update function

File: api/notd/model.py (hunk 7 above)
Comment (Femi-Ogunkola) [reply api/notd/model.py:24]: We already have a class Token

Comment (krishan711) [reply api/notd/model.py:24]: cool this is good then. Please update the table name in the sql to `tbl_token_metadatas` then (add the s at the end)

File: dbschema/1.tables.sql (hunk 1 above)
Comment (krishan711) [reply dbschema/1.tables.sql:24]: `tbl_token_metadatas`

File: dbschema/1.tables.sql [hunk 9]
```diff
@@ -20,3 +20,22 @@ CREATE INDEX tbl_token_transfers_value ON tbl_token_transfers (value);
...
+    updated_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    registry_address TEXT NOT NULL,
+    token_id TEXT NOT NULL,
+    metadata_url TEXT NOT NULL,
+    image_url TEXT NOT NULL
+    name TEXT NOT NULL,
+    description TEXT NOT NULL,
+    attributes TEXT NOT NULL
+
+
```
Comment (krishan711) [inline dbschema/1.tables.sql:36]: theres still a couple of rogue newlines here

File: dbschema/1.tables.sql (hunk 5 above)
Comment (krishan711) [reply dbschema/1.tables.sql:38]: this name is still wrong

File: dbschema/1.tables.sql (hunk 2 above)
Comment (krishan711) [reply dbschema/1.tables.sql:35]: theres still a couple of rogue newlines here

File: api/notd/store/schema_conversions.py [hunk 10]
```diff
@@ -22,3 +24,16 @@ def token_transfer_from_row(row: Mapping) -> TokenTransfer:
         blockHash=row[TokenTransfersTable.c.blockHash],
         blockDate=row[TokenTransfersTable.c.blockDate].replace(tzinfo=datetime.timezone.utc),
     )
+def token_metadata_from_row(row: Mapping) -> TokenMetadata:
```
Comment (krishan711) [inline api/notd/store/schema_conversions.py:27]: newline between functions

File: dbschema/1.tables.sql [hunk 11]
```diff
@@ -20,3 +20,22 @@ CREATE INDEX tbl_token_transfers_value ON tbl_token_transfers (value);
...
+    id BIGSERIAL PRIMARY KEY,
+    created_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    updated_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
+    registry_address TEXT NOT NULL,
+    token_id TEXT NOT NULL,
+    metadata_url TEXT NOT NULL,
+    image_url TEXT NOT NULL
+    name TEXT NOT NULL,
+    description TEXT NOT NULL,
+    attributes TEXT NOT NULL
```
Comment (krishan711) [inline dbschema/1.tables.sql:34]: JSON

File: api/notd/store/schema.py [hunk 12]
```diff
@@ -17,3 +17,18 @@
...
+TokenMetadataTable = sqlalchemy.Table(
+    'tbl_token_metadatas',
+    metadata,
+    sqlalchemy.Column(key='tokenMetadataId', name='id', type_=sqlalchemy.Integer, autoincrement=True, primary_key=True, nullable=False),
+    sqlalchemy.Column(key='createdDate', name='created_date', type_=sqlalchemy.DateTime, nullable=False),
+    sqlalchemy.Column(key='updatedDate', name='updated_date', type_=sqlalchemy.DateTime, nullable=False),
+    sqlalchemy.Column(key='registryAddress', name='registry_address', type_=sqlalchemy.Text, nullable=False),
+    sqlalchemy.Column(key='tokenId', name='token_id', type_=sqlalchemy.Integer, nullable=False),
+    sqlalchemy.Column(key='metadatUrl', name='metadata_url', type_=sqlalchemy.Text, nullable=False),
+    sqlalchemy.Column(key='imageUrl', name='image_url', type_=sqlalchemy.TEXT, nullable=False),
```
Comment (krishan711) [inline api/notd/store/schema.py:30]: many of these have `nullable=False` but are nullable in the table definition below. please check everything is consistent here.

File: api/notd/store/saver.py [hunk 13]
```diff
@@ -51,3 +56,46 @@ async def create_token_transfer(self, retrievedTokenTransfer: RetrievedTokenTran
             blockHash=retrievedTokenTransfer.blockHash,
             blockDate=retrievedTokenTransfer.blockDate,
         )
+
+    async def create_token_metadata(self, tokenId: int, registryAddress: str, metadataUrl: str, imageUrl: int, name: str, description: str, attributes: str) -> TokenMetadata:
```
Comment (krishan711) [inline api/notd/store/saver.py:60]: same here - many of these are not optional but are nullable in the table definition below. please check everything is consistent here.

File: api/notd/model.py [hunk 14]
```diff
@@ -20,6 +21,18 @@ class RetrievedTokenTransfer:
...
+    createdDate: datetime.datetime
+    updatedDate: datetime.datetime
+    registryAddress: str
+    tokenId: int
+    metdataUrl: str
+    imageUrl: int
```
Comment (krishan711) [inline api/notd/model.py:31]: and here many of these are non-optional but are nullable in the table definition below. please check everything is consistent here.

File: api/notd/store/saver.py (hunk 13 above)
Comment (Femi-Ogunkola) [reply api/notd/store/saver.py:60]: Should I equate to _EMPTY_STRING

File: api/notd/store/saver.py [hunk 15]
```diff
@@ -51,3 +59,46 @@ async def create_token_transfer(self, retrievedTokenTransfer: RetrievedTokenTran
...
+            tokenMetadataId=tokenMetadataId,
+            createdDate=createdDate,
+            updatedDate=updatedDate,
+            registryAddress=registryAddress,
+            tokenId=tokenId,
+            metadataUrl=metadataUrl,
+            imageUrl=imageUrl,
+            name=name,
+            description=description,
+            attributes=attributes)
```
Comment (krishan711) [inline api/notd/store/saver.py:87]:
```suggestion
            attributes=attributes,
        )
```

**Review by krishan711 — APPROVED** [review]

**Pull request merged**